File: backend/app/infrastructure/services/ocr_service.py
"""
OCR сервіс для розпізнавання накладних через Gemini API.

Підтримує:
- Ротацію API ключів при помилці 429 (Too Many Requests)
- Повторні спроби при тимчасових помилках (502, 503, timeout)
- Затримку 15 секунд між запитами після 503 помилки
- Затримку 5 секунд між звичайними запитами (щоб уникнути перевищення лімітів)
- Парсинг JSON з відповіді Gemini
- Логування повної відповіді Gemini для діагностики
"""

import asyncio
import json
import logging
import re
from pathlib import Path

from google import genai
from google.genai import types as genai_types

logger = logging.getLogger(__name__)

# Шлях до файлу з ключами відносно кореня проекту
KEYS_FILE_PATH = Path(__file__).resolve().parent.parent.parent.parent / "keys.txt"

# Затримка між звичайними запитами (секунди)
REQUEST_DELAY_SECONDS = 5

# Затримка після 503 помилки (секунди) — збільшено до 15с
RETRY_AFTER_503_DELAY_SECONDS = 15

# Промпт для Gemini (українською)
# Пріоритет: стовідсоткові збіги по штрих-коду.
# Нас цікавить ТІЛЬКИ ціна з ПДВ (cost_price).
INVOICE_PROMPT = """Ти — асистент для розпізнавання прибуткових накладних.
Проаналізуй зображення накладної та поверни ТІЛЬКИ JSON без додаткового тексту.

Формат JSON:
{
  "document_number": "рядок або null",
  "invoice_date": "рядок у форматі YYYY-MM-DD або null",
  "is_fiscal": true/false,
  "supplier_name": "рядок або null",
  "payment_method": "credit" | "bank_transfer" | "cash" | "other" | null,
  "items": [
    {
      "product_name": "рядок",
      "quantity": число,
      "cost_price": число,
      "barcode": "рядок або null"
    }
  ]
}

Правила:
- document_number: номер накладної (наприклад "ПН-00123")
- invoice_date: дата накладної
- is_fiscal: true якщо накладна фіскальна
- supplier_name: назва постачальника
- payment_method: спосіб оплати (credit - в борг, bank_transfer - перерахунок, cash - готівка, other - інше)
- items: масив товарів з назвою, кількістю та ціною з ПДВ (cost_price)
- cost_price: ціна товару з ПДВ (вартість з ПДВ за одиницю)
- barcode: штрих-код товару (EAN-13, 8-14 цифр). Якщо штрих-код чітко видно на накладній — обов'язково поверни його. Якщо не видно або сумніваєшся — null.
- Не повертай price (ціну без ПДВ) — вона не потрібна
- Якщо якесь поле відсутнє на зображенні, використовуй null
- Якщо товарів немає, поверни порожній масив items"""


class OCRService:
    """Сервіс для розпізнавання накладних через Gemini API."""

    def __init__(self):
        self.api_keys: list[str] = []
        self.current_key_index: int = 0
        self.max_retries: int = 3
        self._load_api_keys()

    def _load_api_keys(self) -> None:
        """
        Завантажує ключі з keys.txt.
        Пропускає коментарі (#) та порожні рядки.
        """
        if not KEYS_FILE_PATH.exists():
            logger.warning(f"Файл ключів не знайдено: {KEYS_FILE_PATH}")
            self.api_keys = []
            return

        try:
            with open(KEYS_FILE_PATH, encoding="utf-8") as f:
                keys = []
                for line in f:
                    stripped = line.strip()
                    # Пропускаємо коментарі та порожні рядки
                    if stripped and not stripped.startswith("#"):
                        keys.append(stripped)

            self.api_keys = keys
            logger.info(f"Завантажено {len(keys)} ключ(ів) Gemini API")
        except Exception as e:
            logger.error(f"Помилка завантаження ключів: {e}")
            self.api_keys = []

    def _get_current_key(self) -> str | None:
        """Повертає поточний ключ або None, якщо ключів немає."""
        if not self.api_keys:
            return None
        if self.current_key_index >= len(self.api_keys):
            return None
        return self.api_keys[self.current_key_index]

    def _rotate_key(self) -> None:
        """Перемикає на наступний ключ."""
        self.current_key_index += 1
        logger.info(
            f"Ротація ключа: перехід до індексу {self.current_key_index} "
            f"(всього ключів: {len(self.api_keys)})"
        )

    def _reset_key_index(self) -> None:
        """Скидає індекс ключа на початок."""
        self.current_key_index = 0

    async def _delay_before_next_request(self, context: str = "", delay: int = REQUEST_DELAY_SECONDS) -> None:
        """
        Чекає вказану кількість секунд перед наступним запитом до Gemini API.
        """
        logger.info(f"Затримка {delay}с перед наступним запитом{context}...")
        await asyncio.sleep(delay)

    def _parse_gemini_response(self, response_text: str) -> dict:
        """
        Парсить відповідь Gemini, витягує JSON.

        Шукає JSON у відповіді за допомогою регулярного виразу,
        оскільки Gemini може додавати текст до/після JSON.
        """
        if not response_text:
            raise ValueError("Порожня відповідь від Gemini")

        # Спроба знайти JSON у відповіді (між { та })
        json_match = re.search(r"\{.*\}", response_text, re.DOTALL)
        if not json_match:
            raise ValueError("Не вдалося знайти JSON у відповіді Gemini")

        json_str = json_match.group(0)

        try:
            data = json.loads(json_str)
        except json.JSONDecodeError as e:
            raise ValueError(f"Невалідний JSON у відповіді Gemini: {e}")

        # Валідація обов'язкових полів
        if not isinstance(data, dict):
            raise ValueError("Відповідь Gemini не є об'єктом JSON")

        # Переконуємось, що items — це список
        if "items" not in data or not isinstance(data.get("items"), list):
            data["items"] = []

        return data

    async def analyze_invoice_image(self, image_data: bytes) -> dict:
        """
        Аналізує зображення накладної через Gemini API.

        Args:
            image_data: Байти зображення (PNG, JPEG, WEBP тощо)

        Returns:
            dict: Розпізнані дані накладної у форматі:
                {
                    "document_number": str | None,
                    "invoice_date": str | None,
                    "is_fiscal": bool,
                    "supplier_name": str | None,
                    "payment_method": str | None,
                    "items": list[dict]
                }

        Raises:
            RuntimeError: Якщо всі ключі вичерпано або сталася критична помилка
        """
        exhausted_keys: set[str] = set()
        last_error = None
        first_request = True

        while True:
            api_key = self._get_current_key()

            # Якщо поточний ключ вичерпано — переходимо до наступного
            if api_key is not None and api_key in exhausted_keys:
                self._rotate_key()
                continue

            if api_key is None:
                # Всі ключі вичерпано
                error_msg = (
                    "Всі ключі Gemini API вичерпано (всі повернули 429). "
                    "Додайте нові ключі в keys.txt"
                )
                logger.error(error_msg)
                raise RuntimeError(error_msg)

            for attempt in range(1, self.max_retries + 1):
                try:
                    # Затримка перед кожним запитом, крім найпершого
                    if not first_request:
                        await self._delay_before_next_request(
                            f" (ключ #{self.current_key_index + 1}, спроба {attempt}/{self.max_retries})"
                        )
                    first_request = False

                    logger.info(
                        f"Відправка запиту до Gemini API "
                        f"(ключ #{self.current_key_index + 1}, спроба {attempt}/{self.max_retries})"
                    )

                    client = genai.Client(api_key=api_key)

                    response = client.models.generate_content(
                        model="gemini-3.5-flash",
                        contents=[
                            INVOICE_PROMPT,
                            genai_types.Part.from_bytes(
                                data=image_data,
                                mime_type="image/jpeg",
                            ),
                        ],
                    )

                    # Перевіряємо, чи є текст у відповіді
                    if not response.candidates:
                        raise ValueError("Відповідь Gemini не містить кандидатів")

                    response_text = response.text
                    if not response_text:
                        raise ValueError("Порожній текст у відповіді Gemini")

                    # Логуємо ПОВНУ відповідь Gemini для діагностики
                    logger.info("=== ПОВНА ВІДПОВІДЬ GEMINI (сира) ===")
                    logger.info(response_text)
                    logger.info("=== КІНЕЦЬ ВІДПОВІДІ GEMINI ===")

                    # Парсимо JSON з відповіді
                    result = self._parse_gemini_response(response_text)

                    # Логуємо розпарсену відповідь
                    logger.info("=== РОЗПАРСЕНА ВІДПОВІДЬ GEMINI ===")
                    logger.info(json.dumps(result, ensure_ascii=False, indent=2))
                    logger.info("=== КІНЕЦЬ РОЗПАРСЕНОЇ ВІДПОВІДІ ===")

                    logger.info("Успішно отримано та розпарсено відповідь Gemini")
                    return result

                except Exception as e:
                    error_str = str(e).lower()
                    last_error = e

                    # Логуємо деталі помилки
                    logger.error(f"Помилка Gemini API: {e}")
                    logger.error(f"Тип помилки: {type(e).__name__}")

                    # Перевіряємо, чи це помилка 429 (Too Many Requests)
                    if "429" in error_str or "too many requests" in error_str or "rate limit" in error_str:
                        logger.warning(
                            f"Помилка 429 (Too Many Requests) для ключа #{self.current_key_index + 1}: {e}"
                        )
                        # Додаємо ключ до множини вичерпаних
                        exhausted_keys.add(api_key)
                        # Затримка перед переходом до наступного ключа
                        await self._delay_before_next_request(
                            " (ротація ключа після 429)"
                        )
                        self._rotate_key()
                        break  # Виходимо з циклу спроб, переходимо до наступного ключа

                    # Перевіряємо, чи це помилка 503 (Service Unavailable)
                    is_503 = "503" in error_str or "service unavailable" in error_str

                    # Перевіряємо, чи це тимчасова помилка (502, timeout)
                    is_retryable = any(
                        code in error_str
                        for code in ["502", "timeout", "deadline", "unavailable"]
                    )

                    if is_503:
                        # Для 503 використовуємо збільшену затримку 15с
                        if attempt < self.max_retries:
                            logger.warning(
                                f"Помилка 503 (Service Unavailable) (спроба {attempt}/{self.max_retries}): {e}. "
                                f"Затримка {RETRY_AFTER_503_DELAY_SECONDS}с перед повторною спробою..."
                            )
                            await self._delay_before_next_request(
                                f" (повторна спроба після 503, затримка {RETRY_AFTER_503_DELAY_SECONDS}с)",
                                delay=RETRY_AFTER_503_DELAY_SECONDS
                            )
                            continue  # Повторюємо спробу з тим самим ключем
                        else:
                            logger.error(
                                f"Всі {self.max_retries} спроб вичерпано для ключа #{self.current_key_index + 1} після 503: {e}"
                            )
                            # Затримка перед переходом до наступного ключа
                            await self._delay_before_next_request(
                                " (ротація ключа після вичерпання спроб через 503)"
                            )
                            self._rotate_key()
                            break

                    elif is_retryable and attempt < self.max_retries:
                        logger.warning(
                            f"Тимчасова помилка (спроба {attempt}/{self.max_retries}): {e}. "
                            f"Затримка {REQUEST_DELAY_SECONDS}с перед повторною спробою..."
                        )
                        await self._delay_before_next_request(
                            " (повторна спроба після тимчасової помилки)"
                        )
                        continue  # Повторюємо спробу з тим самим ключем
                    elif is_retryable and attempt >= self.max_retries:
                        logger.error(
                            f"Всі {self.max_retries} спроб вичерпано для ключа #{self.current_key_index + 1}: {e}"
                        )
                        # Затримка перед переходом до наступного ключа
                        await self._delay_before_next_request(
                            " (ротація ключа після вичерпання спроб)"
                        )
                        self._rotate_key()
                        break
                    else:
                        # Інша помилка — не retryable, пробуємо наступний ключ
                        logger.error(
                            f"Помилка Gemini API для ключа #{self.current_key_index + 1}: {e}"
                        )
                        # Затримка перед переходом до наступного ключа
                        await self._delay_before_next_request(
                            " (ротація ключа після помилки)"
                        )
                        self._rotate_key()
                        break

        # Якщо дійшли сюди — всі ключі та спроби вичерпано
        error_msg = f"Не вдалося отримати відповідь від Gemini API: {last_error}"
        logger.error(error_msg)
        raise RuntimeError(error_msg)

File: frontend/src-tauri/crates/torgashka-ocr/src/ocr_service.rs
//! OCR-сервіс — 1:1 Python `OCRService` (ocr_service.py).
//!
//! `analyze_invoice_image`: цикл з ротацією ключів (429 → exhausted),
//! ретраями (503 → 15с, 502/timeout → 5с, max 3 спроби на ключ),
//! затримкою 5с між запитами та парсингом JSON з відповіді Gemini.

use std::collections::HashSet;

use regex::Regex;
use thiserror::Error;

use crate::gemini::{
    classify_error, GeminiClient, RetryKind, REQUEST_DELAY_SECONDS, RETRY_AFTER_503_DELAY_SECONDS,
};
use crate::models::InvoiceData;

/// Помилка OCR-сервісу.
///
/// [`Runtime`] відповідає Python `RuntimeError` (роут повертає `error: msg`),
/// [`Parse`] — Python `ValueError`/інші (`error: "Внутрішня помилка сервера: ..."`).
#[derive(Debug, Error)]
pub enum OcrError {
    /// 1:1 Python `RuntimeError` (вичерпано ключі / фінальна помилка).
    #[error("{0}")]
    Runtime(String),
    /// 1:1 Python `ValueError` та інші винятки (парсинг відповіді).
    #[error("{0}")]
    Parse(String),
}

impl OcrError {
    pub(crate) fn parse(msg: impl Into<String>) -> Self {
        OcrError::Parse(msg.into())
    }
}

/// Regex `(?s)\{.*\}` — inline-прапорець `s` (single-line) = Python `re.DOTALL`:
/// крапка матчить `\n`, тому багаторядковий JSON від Gemini знаходиться цілком.
fn json_regex() -> Regex {
    Regex::new(r"(?s)\{.*\}").expect("regex")
}

/// Обрізає текст до `max` символів (додає "…" якщо обрізано).
fn truncate_chars(s: &str, max: usize) -> String {
    let chars: Vec<char> = s.chars().collect();
    if chars.len() <= max {
        s.to_string()
    } else {
        let mut out: String = chars[..max].iter().collect();
        out.push('…');
        out
    }
}

/// Парсить відповідь Gemini — 1:1 Python `_parse_gemini_response`.
///
/// Діагностика: при будь-якій помилці парсингу (окрім порожньої відповіді)
/// сирий текст Gemini логується через `tracing::error!` (обрізаний до ~2000
/// символів) і перші ~300 символів включаються в повідомлення помилки —
/// фронтенд показує користувачу причину.
pub fn parse_gemini_response(response_text: &str) -> Result<InvoiceData, OcrError> {
    if response_text.is_empty() {
        return Err(OcrError::parse("Порожня відповідь від Gemini"));
    }
    // Будує помилку парсингу: логує сирий текст + додає снипет у повідомлення.
    let parse_err = |context: &str| -> OcrError {
        tracing::error!(
            "Сирий текст відповіді Gemini: {:?}",
            truncate_chars(response_text, 2000)
        );
        OcrError::parse(format!(
            "{context}. Сирий текст відповіді Gemini: {}",
            truncate_chars(response_text, 300)
        ))
    };
    let json_match = json_regex()
        .find(response_text)
        .ok_or_else(|| parse_err("Не вдалося знайти JSON у відповіді Gemini"))?;
    let v: serde_json::Value = serde_json::from_str(json_match.as_str())
        .map_err(|e| parse_err(&format!("Невалідний JSON у відповіді Gemini: {e}")))?;
    if v.as_object().is_none() {
        return Err(parse_err("Відповідь Gemini не є об'єктом JSON"));
    }
    // Python: "items" not in data or not isinstance(items, list) → data["items"] = []
    // Замінюємо items на [] ПЕРЕД десереалізацією (інакше serde падає на не-list).
    let mut v = v;
    if !v.get("items").map(|i| i.is_array()).unwrap_or(false) {
        if let Some(obj) = v.as_object_mut() {
            obj.insert("items".to_string(), serde_json::Value::Array(Vec::new()));
        }
    }
    let data: InvoiceData = serde_json::from_value(v)
        .map_err(|e| parse_err(&format!("Невалідний JSON у відповіді Gemini: {e}")))?;
    Ok(data)
}

/// Сервіс розпізнавання накладних через Gemini — 1:1 Python `OCRService`.
pub struct OcrService {
    client: GeminiClient,
}

impl Default for OcrService {
    fn default() -> Self {
        Self::new()
    }
}

impl OcrService {
    pub fn new() -> Self {
        Self {
            client: GeminiClient::new(),
        }
    }

    /// Внутрішній доступ до клієнта (для тестів).
    pub fn client(&self) -> &GeminiClient {
        &self.client
    }

    /// Аналізує зображення накладної — 1:1 Python `analyze_invoice_image`.
    ///
    /// `content_type` — реальний MIME-тип файлу (вже валідований через
    /// ALLOWED_IMAGE_TYPES в API-роуті); передається в Gemini, щоб модель
    /// отримувала правильний заголовок для PNG/WebP/TIFF/BMP.
    pub async fn analyze_invoice_image(
        &self,
        content_type: &str,
        image_data: &[u8],
    ) -> Result<InvoiceData, OcrError> {
        let mut exhausted_keys: HashSet<String> = HashSet::new();
        let mut first_request = true;

        loop {
            let api_key = self.client.current_key();
            // Якщо поточний ключ вичерпано — переходимо до наступного.
            if let Some(k) = &api_key {
                if exhausted_keys.contains(k) {
                    self.client.rotate();
                    continue;
                }
            }
            let Some(api_key) = api_key else {
                // Всі ключі вичерпано.
                return Err(OcrError::Runtime(
                    "Всі ключі Gemini API вичерпано (всі повернули 429). Додайте нові ключі в keys.txt"
                        .to_string(),
                ));
            };

            for attempt in 1..=self.client.max_retries {
                // Затримка перед кожним запитом, крім найпершого.
                if !first_request {
                    tokio::time::sleep(std::time::Duration::from_secs(REQUEST_DELAY_SECONDS)).await;
                }
                first_request = false;

                match self.client.generate_content(content_type, image_data).await {
                    Ok(response_text) => {
                        let result = parse_gemini_response(&response_text)?;
                        tracing::info!("Успішно отримано та розпарсено відповідь Gemini");
                        return Ok(result);
                    }
                    Err(e) => {
                        let error_str = e.to_string();
                        tracing::error!("Помилка Gemini API: {error_str}");
                        match classify_error(&error_str) {
                            RetryKind::RateLimit => {
                                tracing::warn!("Помилка 429 (Too Many Requests) для ключа");
                                exhausted_keys.insert(api_key.clone());
                                tokio::time::sleep(std::time::Duration::from_secs(
                                    REQUEST_DELAY_SECONDS,
                                ))
                                .await;
                                self.client.rotate();
                                break;
                            }
                            RetryKind::Unavailable503 => {
                                if attempt < self.client.max_retries {
                                    tracing::warn!(
                                        "Помилка 503 (Service Unavailable) (спроба {attempt}/{})",
                                        self.client.max_retries
                                    );
                                    tokio::time::sleep(std::time::Duration::from_secs(
                                        RETRY_AFTER_503_DELAY_SECONDS,
                                    ))
                                    .await;
                                    continue;
                                } else {
                                    tracing::error!("Всі спроби вичерпано після 503");
                                    tokio::time::sleep(std::time::Duration::from_secs(
                                        REQUEST_DELAY_SECONDS,
                                    ))
                                    .await;
                                    self.client.rotate();
                                    break;
                                }
                            }
                            RetryKind::Retryable => {
                                if attempt < self.client.max_retries {
                                    tracing::warn!(
                                        "Тимчасова помилка (спроба {attempt}/{})",
                                        self.client.max_retries
                                    );
                                    tokio::time::sleep(std::time::Duration::from_secs(
                                        REQUEST_DELAY_SECONDS,
                                    ))
                                    .await;
                                    continue;
                                } else {
                                    tracing::error!(
                                        "Всі спроби вичерпано після тимчасової помилки"
                                    );
                                    tokio::time::sleep(std::time::Duration::from_secs(
                                        REQUEST_DELAY_SECONDS,
                                    ))
                                    .await;
                                    self.client.rotate();
                                    break;
                                }
                            }
                            RetryKind::Other => {
                                tracing::error!("Помилка Gemini API, ротація ключа");
                                tokio::time::sleep(std::time::Duration::from_secs(
                                    REQUEST_DELAY_SECONDS,
                                ))
                                .await;
                                self.client.rotate();
                                break;
                            }
                        }
                    }
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_full_json() {
        let text = r#"{"document_number":"ПН-00123","invoice_date":"2026-07-23","is_fiscal":false,"supplier_name":"ТОВ Постачальник","payment_method":"credit","items":[{"product_name":"Молоко","quantity":2,"cost_price":45.5,"barcode":"4820000000001"}]}"#;
        let d = parse_gemini_response(text).unwrap();
        assert_eq!(d.document_number.as_deref(), Some("ПН-00123"));
        assert_eq!(d.invoice_date.as_deref(), Some("2026-07-23"));
        assert!(!d.is_fiscal);
        assert_eq!(d.items.len(), 1);
        assert_eq!(d.items[0].barcode.as_deref(), Some("4820000000001"));
        assert_eq!(d.items[0].quantity, 2.0);
    }

    #[test]
    fn parse_with_surrounding_text() {
        let text = "Ось JSON:\n```json\n{\"document_number\":\"ПН-1\",\"items\":[]}\n```";
        let d = parse_gemini_response(text).unwrap();
        assert_eq!(d.document_number.as_deref(), Some("ПН-1"));
        assert!(d.items.is_empty());
    }

    #[test]
    fn parse_items_not_list_becomes_empty() {
        let text = r#"{"document_number":"x","items":"не список"}"#;
        let d = parse_gemini_response(text).unwrap();
        assert!(d.items.is_empty());
    }

    #[test]
    fn parse_no_items_key() {
        let text = r#"{"document_number":"x"}"#;
        let d = parse_gemini_response(text).unwrap();
        assert!(d.items.is_empty());
        assert!(!d.is_fiscal);
    }

    #[test]
    fn parse_empty_text() {
        assert!(parse_gemini_response("").is_err());
    }

    #[test]
    fn parse_no_json() {
        assert!(parse_gemini_response("просто текст").is_err());
    }

    // ─── Регресія: багаторядковий JSON (як реальна відповідь Gemini) ────

    #[test]
    fn parse_multiline_json_with_fences() {
        let text = "```json\n{\n  \"document_number\": \"ПН-00999\",\n  \"invoice_date\": \"2026-08-19\",\n  \"is_fiscal\": false,\n  \"supplier_name\": \"ТОВ Тест\",\n  \"payment_method\": \"cash\",\n  \"items\": [\n    {\"product_name\": \"Молоко\", \"quantity\": 10, \"cost_price\": 45.5, \"barcode\": \"4820000000001\"}\n  ]\n}\n```";
        let d = parse_gemini_response(text).unwrap();
        assert_eq!(d.document_number.as_deref(), Some("ПН-00999"));
        assert_eq!(d.invoice_date.as_deref(), Some("2026-08-19"));
        assert!(!d.is_fiscal);
        assert_eq!(d.supplier_name.as_deref(), Some("ТОВ Тест"));
        assert_eq!(d.payment_method.as_deref(), Some("cash"));
        assert_eq!(d.items.len(), 1);
        assert_eq!(d.items[0].product_name, "Молоко");
        assert_eq!(d.items[0].quantity, 10.0);
        assert_eq!(d.items[0].barcode.as_deref(), Some("4820000000001"));
    }

    #[test]
    fn parse_multiline_json_no_fences() {
        // Без ```json огорожі — сирий багаторядковий JSON з переносами.
        let text = "{\n  \"document_number\": \"ПН-001\",\n  \"items\": []\n}";
        let d = parse_gemini_response(text).unwrap();
        assert_eq!(d.document_number.as_deref(), Some("ПН-001"));
        assert!(d.items.is_empty());
    }

    #[test]
    fn parse_invalid_json() {
        assert!(parse_gemini_response(r#"{"a": не json}"#).is_err());
    }

    #[test]
    fn parse_not_object() {
        assert!(parse_gemini_response("[1,2,3]").is_err());
    }

    // ─── Діагностика: снипет сирого тексту у повідомленні ────────────────

    #[test]
    fn parse_no_json_message_includes_snippet() {
        let err = parse_gemini_response(
            "Вибачте, я не можу розпізнати зображення. Будь ласка, спробуйте ще раз.",
        )
        .unwrap_err();
        let msg = err.to_string();
        assert!(
            msg.contains("Не вдалося знайти JSON у відповіді Gemini"),
            "{msg}"
        );
        assert!(
            msg.contains("Вибачте, я не можу розпізнати"),
            "повідомлення має містити сирий текст: {msg}"
        );
    }

    #[test]
    fn parse_invalid_json_message_includes_snippet() {
        let err = parse_gemini_response(r#"{"a": не json} щось"#).unwrap_err();
        let msg = err.to_string();
        assert!(msg.contains("Невалідний JSON"), "{msg}");
        assert!(msg.contains("не json"), "{msg}");
    }

    #[test]
    fn parse_empty_message_no_snippet() {
        let err = parse_gemini_response("").unwrap_err();
        assert_eq!(err.to_string(), "Порожня відповідь від Gemini");
    }

    #[test]
    fn truncate_long_text() {
        let long = "а".repeat(5000);
        let t = truncate_chars(&long, 2000);
        assert_eq!(t.chars().count(), 2001); // 2000 + "…"
        assert!(t.ends_with('…'));
        let short = truncate_chars("короткий", 2000);
        assert_eq!(short, "короткий");
    }
}
